com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java

```diff
@@ -240,56 +265,384 @@ public HotCodeReplaceEvent(String message) {
     public void initialize(IDebugAdapterContext context, Map<String, Object> options) {
         // Listen to the built file events.
         ResourcesPlugin.getWorkspace().addResourceChangeListener(this, IResourceChangeEvent.POST_BUILD);
+        this.context = context;
+        currentDebugSession = context.getDebugSession();
 
         // TODO: Change IProvider interface for shutdown event
     }
 
     @Override
     public void resourceChanged(IResourceChangeEvent event) {
-        synchronized (needHotCodeReplace) {
-            needHotCodeReplace.set(true);
-        }
-
         if (event.getType() == IResourceChangeEvent.POST_BUILD) {
             ChangedClassFilesVisitor visitor = getChangedClassFiles(event);
             if (visitor != null) {
                 List<IResource> resources = visitor.getChangedClassFiles();
                 List<String> fullyQualifiedName = visitor.getQualifiedNamesList();
                 if (resources != null && !resources.isEmpty()) {
                     doHotCodeReplace(resources, fullyQualifiedName);
-                } else {
-                    synchronized (needHotCodeReplace) {
-                        needHotCodeReplace.set(false);
-                        needHotCodeReplace.notifyAll();
-                    }
                 }
             }
         }
     }
 
     @Override
-    public CompletableFuture<List<String>> redefineClasses() {
-        return CompletableFuture.supplyAsync(() -> {
-            synchronized (needHotCodeReplace) {
+    public void redefineClasses(Consumer<List<String>> consumer) {
+        this.consumer = consumer;
```

Comment:
The event may take multiple subscribers and allow unsubscribing. RXJava interface is a good reference.